File: src/_215.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        return reverse_findk(nums,0,nums.size() - 1,k);
    }
    int reverse_findk(vector<int> &nums, int left, int right, int k)
    {
        int i, j;
        i = j = left;
        int pivot = nums[right];
        while (j < right)
        {
            if(nums[j] < pivot)
            {
                swap(nums[i],nums[j]);
                ++i;
            }
            ++j;
        }
        swap(nums[i],nums[right]);
        if(right - i + 1 == k)
        {
            return nums[i];
        }
        else if(right - i + 1 > k)
        {
            return reverse_findk(nums,i+1,right,k);
        }
        else
        {
            return reverse_findk(nums,left,i-1,k - right + i -1);
        }
    }
    inline void swap(int & a,int & b)
    {
        int temp = b;
        b = a;
        a = temp;
    }
};

int main()
{
    Solution test;
    vector<int> a = {3,2,1,5,6,4};
    cout<<test.findKthLargest(a,2)<<endl;
    return 0;
}

File: 215.py
from typing import List
from random import randint

class Solution:
    def partition(self, nums: List[int], left: int, right: int):
        i = randint(left, right)
        pivot = nums[i]
        nums[left], nums[i] = nums[i], nums[left]

        i, j = left + 1, right
        while True:
            while i <= j and nums[i] < pivot:
                i += 1
            while i <= j and nums[j] > pivot:
                j -= 1
            if i >= j:
                break
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1
        nums[left], nums[j] = nums[j], nums[left]
        return j


    def findKthLargest(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left, right = 0, n - 1
        
        while True:
            j = self.partition(nums, left, right)
            if j == n - k:
                return nums[j]
            elif j > n - k:
                right = j - 1
            else:
                left = j + 1